hyp/base.py

```diff
@@ -32,15 +33,25 @@ def _respond(self, instance_or_instances, meta=None, links=None, linked=None):
         if meta is not None:
             document['meta'] = self.build_root_meta(meta)
 
-        if links is not None:
-            document['links'] = self.build_root_links(links)
+        collector = Collector()
 
         if linked is not None:
-            document['linked'] = self.build_root_linked(linked)
+            self.build_root_linked(collector, linked)
+
+        if links is not None:
+            self.build_root_links(collector, links)
+
+        if collect:
+            links = self.LINKS.keys()
+            document[self.TYPE] = self.build_resources(instance_or_instances, links, collector)
+        else:
+            document[self.TYPE] = self.build_resources(instance_or_instances, links)
 
-        document[self.TYPE] = self.build_resources(instance_or_instances, links)
+        document['linked'] = collector.get_linked_dict()
+        document['links'] = collector.get_links_dict()
 
-        return document
+        # Filter out empty lists
```

Comment:
Anything in tests/test_resource.py requires it.

It used to be that if linked or links was unspecified, the condition block would skip over adding the elements to the document. This is no longer the case as collector is used and multiple conditions can cause data to be present. Filtering them out at the end is a single line compared to having to add it as a check for each element individually.